File: scala/dllib/src/main/scala/com/intel/analytics/bigdl/dllib/utils/tf/loaders/Slice.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intel.analytics.bigdl.utils.tf.loaders

import java.nio.ByteOrder

import com.intel.analytics.bigdl.Module
import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
import com.intel.analytics.bigdl.nn.ops.{Slice => SliceOps}
import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils.tf.Context
import org.tensorflow.framework.NodeDef

import scala.reflect.ClassTag

class Slice extends TensorflowOpsLoader {


  override def build[T: ClassTag](nodeDef: NodeDef, byteOrder: ByteOrder,
    context: Context[T])(implicit ev: TensorNumeric[T]): Module[T] = {
    new SliceLoadTF[T]()
  }
}

class SliceLoadTF[T: ClassTag]()(implicit ev: TensorNumeric[T]) extends Adapter[T](Array(2, 3)) {
  import Utils._

  override def build(tensorArrays: Array[Tensor[_]]): AbstractModule[Activity, Activity, T] = {
    val size = tensorArrays(1).asInstanceOf[Tensor[Int]]
    SliceOps[T](toArray(tensorArrays(0).asInstanceOf[Tensor[Int]]),
      toArray(tensorArrays(1).asInstanceOf[Tensor[Int]]))
  }
}

File: pyzoo/zoo/pipeline/api/onnx/mapper/slice.py
# -*- coding: utf-8 -*
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from zoo.pipeline.api.onnx.mapper.operator_mapper import OperatorMapper
import zoo.pipeline.api.keras.layers as zlayers
import zoo.pipeline.api.autograd as autograd
import bigdl.nn.layer as blayer


class SliceMapper(OperatorMapper):
    def __init__(self, node, _params, _all_tensors):
        super(SliceMapper, self).__init__(node, _params, _all_tensors)

    def _extract_model_inputs(self):
        """
        :return: list of OnnxInput
        """
        return [self._to_zoo_input(self._input_list[0])]

    def _to_tensor(self):
        lens = []
        input = self.model_inputs[0].zvalue
        ends = self.onnx_attr['ends']
        starts = self.onnx_attr['starts']
        if "axes" in self.onnx_attr.keys():
            axes = self.onnx_attr['axes']
        else:
            axes = range(len(starts))
        for j in range(len(starts)):
            lens.append(ends[j] - starts[j])
            # slice(, len=-1) equals to slice(, len=length)
            # y = x[:2,0:-1] means start is(0,0) and ends is(2,-1)
            # which is equivalent to slice(,len=-2) as "end=-1" is exclusive here.
            if lens[j] < 0:
                lens[j] -= 1
        for i in range(len(starts)):
            input = input.slice(dim=int(axes[i]), start_index=int(starts[i]), length=int(lens[i]))
        return input
